### PR TITLE
Alyapunov/gh 5628 fix bug in tx manager gap

### LINKED ISSUE #5628: Fix bugs in memtx TX manager
This is a cumulative ticket.
There's a lot of problems that happen because TX manager cannot track read when it didn't find anything, or range read. We need to grab interval trees from vinyl to track those read.

### PR DESCRIPTION
**Pull request by alyapunov**
After implementation of point hole trackers only one problem has
left. When a transaction reads a value from a TREE index by key
that key may be partial and/or search type may be GE, GT etc. In
this case point hole hash table doesn't provide track of writes
into some key interval.

Implement gap trackers for those purposes. The main idea is to
store those trackers in a story of a closest tuple to the right of
the lookup (aka successor) or directly in index if there no
successor (for example if the space in empty).

Closes #5628

### CODE REVIEW

File: src/box/memtx_tx.c [hunk 1]
```diff
@@ -1644,6 +1797,85 @@ memtx_tx_track_point_slow(struct txn *txn, struct space *space,
...
+ * when interation type is not EQ or when the key is not full (otherwise
+ * it's faster to use memtx_tx_track_point).
+ * @return 0 on success, -1 on memory error.
+ */
+int
+memtx_tx_track_gap_slow(struct txn *txn, struct space *space, uint32_t index,
+			struct tuple *successor, enum iterator_type type,
+			const char *key, uint32_t part_count)
+{
+	if (txn->status != TXN_INPROGRESS)
```
Comment (Korablev77) [inline src/box/memtx_tx.c:1850]: I'd move this check to memtx_tx_track_gap() as well

Comment (alyapunov) [reply src/box/memtx_tx.c:1850]: In `memtx_tx.h`? I thought about that, but this add a dependency from `txn.h`. Are you sure that it is worth it?

File: test/box/tx_man_intervals.test.lua [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+s:delete{3}
+
+tx1:begin()
+tx2:begin()
+tx1('s:select{}')
+tx2('s:select{}')
+tx1('s:replace{1, 1}')
+tx2('s:replace{2, 2}')
+tx1:commit()
+tx2:commit()
```
Comment (EvgenyMekhanik) [inline test/box/tx_man_intervals.test.lua:54]: I try to do this test case for 'hash' index type and got other results:
os.execute("rm -rf *.snap *.xlog")

box.cfg{memtx_use_mvcc_engine  = true}

txn_proxy = require('txn_proxy')

tx1 = txn_proxy.new()
tx2 = txn_proxy.new()

s = box.schema.space.create('test')
s:format({
         {name = 'id', type = 'unsigned'},
         })

i = s:create_index('pk', {type = 'tree', parts = {'id'}})

s:replace{3, 3}

tx1:begin()
tx2:begin()
tx1('s:select{}')
tx2('s:select{}')
tx1('s:replace{1, 1}')
tx2('s:replace{2, 2}')
tx1:commit()
tx2:commit()

For type == tree tx2 fails (it's should be so), but for type == hash not!
As i see now this patch fix problem only for `tree` index type!

**Review by Korablev77 — APPROVED** [review]
LGTM

**Pull request merged**